File: ios/TrueImageThumbnails.m
#import "TrueImageThumbnails.h"

@implementation TrueImageThumbnails {
  NSCache<NSString *, id> *_cache;
  dispatch_queue_t _queue;
}

+ (TrueImageThumbnails *)shared
{
  static TrueImageThumbnails *shared;
  static dispatch_once_t once;
  dispatch_once(&once, ^{
    shared = [TrueImageThumbnails new];
  });
  return shared;
}

- (instancetype)init
{
  if (self = [super init]) {
    _cache = [NSCache new];
    _cache.countLimit = 64;
    _queue = dispatch_queue_create("TrueImage.thumbnails", DISPATCH_QUEUE_SERIAL);
    dispatch_set_target_queue(_queue, dispatch_get_global_queue(QOS_CLASS_USER_INITIATED, 0));
  }
  return self;
}

- (CGImageRef)cachedForKey:(NSString *)key
{
  return (__bridge CGImageRef)[_cache objectForKey:key];
}

- (void)makeFromImage:(UIImage *)image
            pixelSize:(CGSize)pixelSize
                  key:(NSString *)key
           completion:(void (^)(CGImageRef))completion
{
  CGImageRef hit = [self cachedForKey:key];
  if (hit) {
    completion(hit);
    return;
  }
  NSCache *cache = _cache;
  dispatch_async(_queue, ^{
    UIGraphicsImageRendererFormat *format = [UIGraphicsImageRendererFormat new];
    format.scale = 1;
    format.opaque = NO;
    UIGraphicsImageRenderer *renderer = [[UIGraphicsImageRenderer alloc] initWithSize:pixelSize format:format];
    UIImage *thumb = [renderer imageWithActions:^(UIGraphicsImageRendererContext *_) {
      [image drawInRect:CGRectMake(0, 0, pixelSize.width, pixelSize.height)];
    }];
    CGImageRef cg = thumb.CGImage;
    if (cg) {
      [cache setObject:(__bridge id)cg forKey:key];
    }
    dispatch_async(dispatch_get_main_queue(), ^{
      completion(cg);
    });
  });
}

@end

File: ios/TrueImageThumbnails.swift
import UIKit

/// Resamples a decoded image once to the size it is actually drawn at when
/// the source is more than twice as large, so a 1200 px cover in a 60 pt
/// row does not cost the GPU a full-size texture every frame.
final class TrueImageThumbnails {
  static let shared = TrueImageThumbnails()

  private let cache: NSCache<NSString, CGImage> = {
    let cache = NSCache<NSString, CGImage>()
    cache.countLimit = 64
    return cache
  }()

  private let queue = DispatchQueue(label: "TrueImage.thumbnails", qos: .userInitiated)

  func cached(key: String) -> CGImage? {
    cache.object(forKey: key as NSString)
  }

  /// Renders off the main thread and calls back on it.
  func make(
    from image: UIImage,
    pixelSize: CGSize,
    key: String,
    completion: @escaping (CGImage?) -> Void
  ) {
    if let hit = cached(key: key) {
      completion(hit)
      return
    }
    queue.async { [cache] in
      let format = UIGraphicsImageRendererFormat()
      format.scale = 1
      format.opaque = false
      let renderer = UIGraphicsImageRenderer(size: pixelSize, format: format)
      let thumb = renderer.image { _ in
        image.draw(in: CGRect(origin: .zero, size: pixelSize))
      }.cgImage
      if let thumb { cache.setObject(thumb, forKey: key as NSString) }
      DispatchQueue.main.async { completion(thumb) }
    }
  }
}
